### PR TITLE
Enhanced code splitting

### PR DESCRIPTION
**Pull request by mattgperry**
Improvement for https://github.com/framer/motion/issues/254

This PR refactors the animation controls to make them more code-splittable, preferring individual functions.

Currently brings overall bundle down to 27kb and brings the `m` component import down to under 12.5kb.

### CODE REVIEW

File: src/animation/AnimationControls.ts [hunk 1]
```diff
@@ -42,42 +38,7 @@ export class AnimationControls {
      *
      * @internal
      */
-    private componentControls = new Set<VisualElementAnimationControls>()
-
-    /**
-     * A map of variants that can be later referenced via `start(variantLabel)`
```
Comment (mattgperry) [inline src/animation/AnimationControls.ts:48]: This stuff was legacy from pre-1.0

**Review by nvh — COMMENTED** [review]
@mattgperry  As this is quite a bit of changes, do you have any directions on how we can best review this?

**Comment by mattgperry** [comment]
Well the main change is converting `VisualElementAnimationControls` to a bunch of functions that now live in `render/VisualElement/utils` and some changes to context. As a large refactor with passing tests I suppose the main thing is not having done anything outrageously stupid.

**Comment by mattgperry** [comment]
Bug in https://codesandbox.io/s/framer-motion-2-drag-to-reorder-fc4rt

File: src/motion/__tests__/variant.test.tsx [hunk 2]
```diff
@@ -9,519 +9,521 @@ describe("animate prop as variant", () => {
...
-    })
+    // const childVariants: Variants = {
+    //     hidden: { opacity: 0, x: -100, transition: { type: false } },
+    //     visible: { opacity: 1, x: 50, transition: { type: false } },
+    // }
+
+    // test("animates to set variant", async () => {
+    //     const promise = new Promise((resolve) => {
+    //         const x = motionValue(0)
+    //         const onComplete = () => resolve(x.get())
```
Comment (hemlok) [inline src/motion/__tests__/variant.test.tsx:20]: Are these tests replaced somewhere? Are we losing coverage by committing this with them commented out?

File: src/motion/utils/use-variants.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        visualElement.isMounted = true
+        return () => {
+            visualElement.isMounted = false
+            parent?.variantChildren?.delete(visualElement)
+        }
+    }, [])
+
+    /**
+     *
+     */
```
Comment (hemlok) [inline src/motion/utils/use-variants.ts:112]: 👀

File: src/motion/utils/use-variants.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
+import { useContext, useEffect, useMemo } from "react"
```
Comment (hemlok) [inline src/motion/utils/use-variants.ts:1]: Nicely documented file 💪

File: src/motion/__tests__/variant.test.tsx (hunk 2 above)
Comment (mattgperry) [reply src/motion/__tests__/variant.test.tsx]: Comments left by accident, thanks!

**Review by hemlok — APPROVED** [review]
Nothing jumps out as problematic going through the code, but it's a big diff! Going to go through the examples now but preliminarily approving 👍

**Comment by hemlok** [comment]
Is this example broken? `animation-between-value-types`

All other examples seem to be working as expected 🎉

**Pull request merged**